### ISSUE TITLE
Loss NaNs as a function of dataset size

### ISSUE THREAD

**Issue by r-karimi** [ElArkk/jax-unirep#94]
Hello! Thanks for writing this great reimplementation of UniRep in JAX.

I'm running into a puzzling issue with the estimation of the holdout set loss in the fit function. I'm working on the Graham HPC cluster (https://docs.computecanada.ca/wiki/Graham) with 4 NVIDIA t4 chips, 16 GB RAM each. I'm trying to use the fit function on a dataset of about 50,000 sequences, all between 450 and 550 residues, using the random batching method.

I begin by shuffling the sequences, then taking the first 5,000 as holdout sequences and feeding the rest into the fit function, similarly to the workflow suggested in the docs.

When I fit a subset of <=11,000 sequences from the larger dataset of 50,000 sequences, I can set the learning rate to a reasonable 1e-5 and observe the loss dropping over several training epochs. However, as soon as I pass to the 'sequences' argument a character list of more than roughly 11,000 sequences, the evotuning log reports a loss value of NaN. I've found this threshold of ~11,000 sequences empirically.

I know there have been issues in the past with the final softmax layer producing NaNs and negative values, but I can't think of a reason the number of training sequences would factor into that. I would be grateful to hear any ideas or suggestions you might have.

Thanks!

**Comment by r-karimi** [comment]
I did a bit more testing and wanted to give some more information to make troubleshooting easier. I manually checked my sequences for odd characters (e.g. *) that might be throwing things off. I also wanted to see if sorting by length and then fitting smaller subsets of the sequences would help (i.e. sort by length and then fit the 1000 longest sequences), but that also didn't seem to help.

I also noticed that the fit function always seems to calculate a reasonable training and holdout set loss before beginning to fit--it is only after the first epoch and beyond that NaNs appear in the estimates of the training and holdout losses.

If it's of any assistance, I'm running a batch size of 150, since the t4 chips have a GPU RAM of 16 GB.

**Comment by ericmjl** [comment]
Hey @r-karimi!

Firstly, thanks for pinging in! I appreciate the effort you put in to try debugging.

From the description, I can see you tried a bunch of things, though there are still some gaps missing in my head, perhaps arising from the way we structured our code - the average loss is calculated on whatever the batch size is set to, which is going to be less than 11K (or 5K). I can imagine the average loss giving NaNs if it were calculated on all 11K.

NaN errors are tough to debug, and I think the only way I can think of is to go function by function inside the [fit function](https://github.com/ElArkk/jax-unirep/blob/master/jax_unirep/evotuning.py#L149). If you're up for it, we can schedule a call spanning 3 time zones (pacific time, eastern time, and central european time) on a Saturday. I just have to coordinate a bit with my wife to make sure baby care is taken care of. Alternatively, if you want to try your hand at debugging, this is what we would do in a call:

1. Isolate the `fit` function contents inside a script/Jupyter notebook.
2. Run the code line by line to inspect the outputs, and see where the first time NaN shows up. The key culprits I would pay attention to at first are:
    1. Line 336: https://github.com/ElArkk/jax-unirep/blob/master/jax_unirep/evotuning.py#L336
    2. Line 349: https://github.com/ElArkk/jax-unirep/blob/master/jax_unirep/evotuning.py#L349
    3. Line 356: https://github.com/ElArkk/jax-unirep/blob/master/jax_unirep/evotuning.py#L356
3. If that fails to be informative, the second strategy would be to use the utility functions that we have to manually pass the sequences through the model until you hit the loss function calculation, then compute the derivative. This is a bit more involved, but can be much more informative if the cheap/easy steps above fail to be informative.

Let me know if you want to call this Saturday, 19 December; also, if you get some practice debugging according to the steps above, then our call together will be much more productive, as you'll have a bit more mental familiarity with the codebase, even if you don't have an answer by then. The most likely time that will fit all three of us (yourself, myself and @ElArkk) is 10 am PT/1 pm ET/7 pm CET.

**Comment by r-karimi** [comment]
Thanks @ericmjl! I isolated the fit function as you described above. It seems like the NaNs appear after about 10 iterations of the step function within fit--much sooner than one full epoch. This seemed interesting to me. I appreciate the offer to call! I will send you an email to confirm a time. Thank you again for your help.

**Comment by ericmjl** [comment]
Hey @r-karimi! I hope your exams went well. Don't fret them -- I was in the MBIM program at UBC for 1 year (before I switched to ISCI later). Didn't do exactly the best in MBIM, but still somehow made it out in one piece haha.

Anyways, I propose we try for a session next week. Given that it's holiday season, we might all have a bit of time. Does Tuesday or Wednesday sound good? @ElArkk, will you able to join us? (It's going to be late on your end, so don't feel pressured, alternatively feel free to let me know if we should try a different time.)

**Comment by r-karimi** [comment]
Sure, either Tuesday or Wednesday should work for me. Any time after 10 AM would do on my end!

Thanks again for all the help.

**Comment by ElArkk** [comment]
Hi @r-karimi  and @ericmjl ! I'd gladly join tomorrow evening (CET). Wednesday I'm not able to probably. So if tomorrow works out for everybody, great, otherwise feel free to go ahead without me :)

**Comment by r-karimi** [comment]
Hi all, I'm good for 10 AM/1 PM/7 PM in our respective time zones at any point this week!

**Comment by ericmjl** [comment]
@ElArkk @r-karimi I just looped back this morning after our new nanny came in. It's been an exhausting time struggling on our own. If it's not too late for y'all, I've got a [google meet link](https://meet.google.com/fki-wpag-pei) for us all. I'll be up there at 1 pm today. (Skipping the invite since it's so last-minute.) And if I don't see you up at 1 pm ET, @r-karimi, we can kick the can to tomorrow, not a problem.

**Comment by ericmjl** [comment]
Where we left off:

- [ ] Identified (using buf.to_py().shape) that the model params of shapes: (10, 1900), (10,7600), (1900,7600) (7600,) etc. are going to NaN on possibly one of the lax.scan computations.

difficulty: 

- Not easy to isolate the sequence batch that generates the NaN computations.
- Not easy to identify which function exactly gave us NaNs, though grad is the biggest culprit atm.

**Comment by ericmjl** [comment]
@r-karimi this morning I mulled over some of your thoughts on the grad function returning NaN. Looking through the test suite, I saw that we didn't have a NaN unit test for the gradient function of `sigmoid`. As such, I went in and used Hypothesis to find the case in which our "grad" of sigmoid would return a null value -- and promptly found that sigmoid would give a NaN value at values < -89. The power of using Hypothesis to test our numerical code shows up again! I would never have found this out by manual testing, or even auto-manual testing.

The test was easy to write:

```python
"""Tests for activation functions.

In particular, we are looking for tests that cause NaN errors in grads.
"""
from hypothesis import strategies as st, given, settings
from jax_unirep.activations import sigmoid
import pytest
import jax.numpy as np
from jax import grad


@pytest.mark.parametrize("version", ["tanh", "exp"])
@given(x=st.floats(allow_nan=False, allow_infinity=False))
@settings(deadline=None)
def test_sigmoid(x, version):
    """Check for null gradient issues."""
    result = sigmoid(x, version=version)
    assert not np.isnan(result)

    dsigmoid = grad(sigmoid)
    dresult = dsigmoid(x, version=version)
    assert not np.isnan(dresult)
```

On the basis of this, I changed the implementation of the sigmoid function ever so slightly:

```python
def sigmoid(x, version="tanh"):
    """Sigmoid activation function.

    Two versions are provided: "tanh" and "exp".
    "exp" is used in the re-implementation.
    """
    sigmoids = {
        "tanh": lambda x: 0.5 * np.tanh(x) + 0.5,
        "exp": lambda x: safe_sigmoid_exp(x),
    }
    return sigmoids[version](x)


def safe_sigmoid_exp(x, clip_value=-88):
    """
    Safe version of exp version of sigmoid.

    Based on the test "test_sigmoid", we found that a value of -89.0
    gives us NaN values when calculating the gradient of sigmoid.
    As such, we clip -x to a minimum value of -88.0.
    """
    x = np.clip(x, a_min=-88)
    return 1 / (1 + np.exp(-x))
```

I went back to the dataset you uploaded and didn't observe any NaN issues anymore. I think we might be good here. I am pushing up the changes, and will cut a patch release shortly.

**Comment by ElArkk** [comment]
Great digging @ericmjl ! Shows the value of property-based testing :)

**Comment by ericmjl** [comment]
@r-karimi I think we can skip today's pair calling session, btw. Let's sync up asynchronously up here :smile:.

**Comment by r-karimi** [comment]
@ericmjl and @ElArkk, thank you both very much for all of your help. @ericmjl, I'm really glad you were able to find the bug without much issue in the end, and I'm glad it was a straightforward fix. I'll be sure to do some testing with this new patch and let you know if there are any further issues. Thanks again!

**Comment by ericmjl** [comment]
My pleasure, @r-karimi!

**Comment by r-karimi** [comment]
Hi @ericmjl and @ElArkk, the HPC cluster I'm using doesn't seem to be allowing me to install from Github directly. As soon as you push version 2.0.1 to PyPI, I'll be able to test out the new sigmoid activation on my dataset. Do you have an idea of when you're planning on updating the package on PyPI?

**Comment by ericmjl** [comment]
I am hoping @ElArkk gets a few more rounds of cutting releases (practice + confidence in the workflow), but if it's something urgent for you, @r-karimi, happy to handle it tomorrow morning first thing.

**Comment by r-karimi** [comment]
Fair enough! Just wanted to make sure a new release was on the horizon :)

No rush on my end, whatever works best for you two.

**Comment by ElArkk** [comment]
@r-karimi @ericmjl happy new year to you both :) I was off the grid since Christmas, will try cutting a release today or latest tomorrow!

**Comment by ElArkk** [comment]
Done @r-karimi , hope it now works for you, fingers crossed 😄
